### PR TITLE
Update image upgrade code

### PR DESCRIPTION
**Pull request by AmeyShikerkar**
Signed-off-by: Amey Shikerkar <amey.shikerkar@ibm.com>

### CODE REVIEW

**Comment by bpradipt** [comment]
@AmeyShikerkar can you please create an issue for this

**Comment by AmeyShikerkar** [comment]
@bpradipt @yussufsh , please review.

**Comment by AmeyShikerkar** [comment]
@arcprabh

**Comment by AmeyShikerkar** [comment]
Issue https://github.com/ocp-power-automation/ocp4-playbooks/issues/90

**Review by yussufsh — COMMENTED** [review]
Hi @AmeyShikerkar 
I had one question regarding this change. What will happen if I specify both `upgrade_version` and `upgrade_image`?

**Comment by AmeyShikerkar** [comment]
This scenario is yet to be tested. 
What should be the expected behavior? should it upgrade to version and then upgrade to image or give error if both are specified?

**Comment by yussufsh** [comment]
> 
> This scenario is yet to be tested.
> What should be the expected behavior? should it upgrade to version and then upgrade to image or give error if both are specified?

I feel you should give preference to one method. I am sure one upgrade will fail if both are started together.

**Comment by AmeyShikerkar** [comment]
Upgrade Version OCP 4.5.20->OCP 4.5.21 and upgrade to nightly 4.6.6

upgrade_vars.yaml 
```
upgrade_version: "4.5.21"
upgrade_channel: "fast-4.5"
upgrade_image: "quay.io/openshift-release-dev/ocp-release@sha256:2517e14d07fbb1403cfd410911d2c78ac60aa1a71565933310a49c3d72736511"
pause_time: 90
delay_time: 600
```
Tried running playbook with both parameters,
ansible-playbook -i inventory -e @upgrade_vars.yaml playbooks/upgrade.yaml 

```
TASK [Gathering Facts] *************************************************************************************************************************
ok: [9.114.97.51]

TASK [include_role : ocp-upgrade] **************************************************************************************************************

TASK [ocp-upgrade : Set upgrade channel] *******************************************************************************************************
changed: [9.114.97.51]

TASK [ocp-upgrade : Upgrade cluster] ***********************************************************************************************************
changed: [9.114.97.51]

TASK [ocp-upgrade : Waiting until upgrade completes] *******************************************************************************************
Pausing for 5400 seconds
(ctrl+C then 'C' = continue early, ctrl+C then 'A' = abort)
Press 'C' to continue the play or 'A' to abort
ok: [9.114.97.51]

TASK [ocp-upgrade : Check for upgrade completion] **********************************************************************************************
changed: [9.114.97.51]

TASK [ocp-upgrade : Upgrade cluster] ***********************************************************************************************************
changed: [9.114.97.51]

TASK [ocp-upgrade : Waiting until upgrade completes] *******************************************************************************************
Pausing for 5400 seconds
(ctrl+C then 'C' = continue early, ctrl+C then 'A' = abort)
ok: [9.114.97.51]

TASK [ocp-upgrade : Check for upgrade completion] **********************************************************************************************
FAILED - RETRYING: Check for upgrade completion (5 retries left).
FAILED - RETRYING: Check for upgrade completion (4 retries left).
FAILED - RETRYING: Check for upgrade completion (3 retries left).
FAILED - RETRYING: Check for upgrade completion (2 retries left).
FAILED - RETRYING: Check for upgrade completion (1 retries left).
```

Observed that after OCP version upgrade, image upgrade continued although it is in progress
```
]# oc version
Client Version: 4.5.5
Server Version: 4.5.20
Kubernetes Version: v1.18.3+ca4017d

# oc version
Client Version: 4.5.5
Server Version: 4.5.21
Kubernetes Version: v1.18.3+a932adb

# oc get clusterversion
NAME      VERSION   AVAILABLE   PROGRESSING   SINCE   STATUS
version   4.5.21    True        True          138m    Working towards 4.6.6: 11% complete
```
Although playbook failed , cluster image upgrade  complete
```
TASK [ocp-upgrade : Check for upgrade completion] **********************************************************************************************
FAILED - RETRYING: Check for upgrade completion (5 retries left).
FAILED - RETRYING: Check for upgrade completion (4 retries left).
FAILED - RETRYING: Check for upgrade completion (3 retries left).
FAILED - RETRYING: Check for upgrade completion (2 retries left).
FAILED - RETRYING: Check for upgrade completion (1 retries left).
fatal: [9.114.97.51]: FAILED! => {"attempts": 5, "changed": true, "cmd": "oc get clusterversion -o json|jq \".items[0].status.history[0].state\"\n", "delta": "0:00:00.508541", "end": "2020-12-08 05:49:08.593608", "rc": 0, "start": "2020-12-08 05:49:08.085067", "stderr": "", "stderr_lines": [], "stdout": "\"Partial\"", "stdout_lines": ["\"Partial\""]}

PLAY RECAP *************************************************************************************************************************************
9.114.97.51                : ok=7    changed=4    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0

# oc get clusterversion
NAME      VERSION   AVAILABLE   PROGRESSING   SINCE   STATUS
version   4.6.6     True        False         11m     Cluster version is 4.6.6
```

@yussufsh,  upgrade is happening sequentially as seen above. Is this acceptable(We have such scenario to upgrade first to a version and then upgrade to nightly? or we need give preference to one method?

**Comment by AmeyShikerkar** [comment]
As discussed with @yussufsh, upgrade to image should take preference if both upgrade version and  upgrade image are specified. Updated code for release image to take preference over upgrade version.

**Review by yussufsh — DISMISSED** [review]
lgtm

**Comment by Prajyot-Parab** [comment]
lgtm!

**Comment by AmeyShikerkar** [comment]
Thanks @yussufsh , @Prajyot-Parab 
Do we have any more review comments?

**Review by Prajyot-Parab — DISMISSED** [review]

**Comment by arcprabh** [comment]
@bpradipt - pls review.

File: playbooks/roles/ocp-upgrade/README.md [hunk 1]
```diff
@@ -1,7 +1,7 @@
 ocp-upgrade: Upgrade OCP cluster
 =========
 
-This module will upgrade an existing OCP cluster to a higher version.
+This module will upgrade an existing OCP cluster to a higher version or release image. If both upgrade version and upgrade image are specified then cluster will be upgraded to release image.
```
Comment (bpradipt) [inline playbooks/roles/ocp-upgrade/README.md]: This module will upgrade an existing OCP cluster based on `upgrade_version` or `upgrade_image`. If both `upgrade_version` and `upgrade_image` are specified then `upgrade_image` value gets preference.

Comment (AmeyShikerkar) [reply playbooks/roles/ocp-upgrade/README.md]: Incorporated review comments in the doc. Thanks!

**Review by bpradipt — APPROVED** [review]
/lgtm

**Review by yussufsh — APPROVED** [review]
/approve

**Comment by yussufsh** [comment]
Hi @AmeyShikerkar I think this feature is also needed in previous release branches. Could you please help us cherry-pick them and raise PRs?

**Pull request merged**